### PR TITLE
Refactor and setup git ssh keys uniformly everywhere

### PR DESCRIPTION
**Pull request by jaseemabid**
There is honestly no way to test this change thoroughly and a lot of this might
be dead code. I'm hoping to run the planner builds once and merge this if those
builds go green :crossed_fingers:

### CODE REVIEW

File: vars/goRelease.groovy [hunk 1]
```diff
@@ -31,15 +31,11 @@ def call(body) {
...
         git "https://github.com/${ghOrg}/${prj}.git"
 
-        sh "git config user.email fabric8-admin@googlegroups.com"
-        sh "git config user.name fabric8-release"
+        flow.setupGitSSH()
         sh "git remote set-url origin git@github.com:${ghOrg}/${prj}.git"
 
         container(name: 'go') {
             stage ('build binary'){
-                sh 'chmod 600 /root/.ssh-git/ssh-key'
```
Comment (hrishin) [inline vars/goRelease.groovy]: dropping these lines totally?

Comment (hrishin) [reply vars/goRelease.groovy]: Though https://github.com/fabric8io/fabric8-pipeline-library/pull/396/files#diff-0008f0d28f3aed4072c34ae9bbda634aR34 going to make sure

File: vars/npmRelease.groovy [hunk 2]
```diff
@@ -6,27 +6,24 @@ def call(body) {
     body.delegate = config
     body()
 
-    def gitEmail = config.gitEmail ?: 'fabric8-admin@googlegroups.com'
-    def gitUserName = config.gitUserName ?: 'fabric8-release'
     def branch = config.branch
+    def flow = new io.fabric8.Fabric8Commands()
 
-    sh """
-       git config user.email ${gitEmail}
```
Comment (hrishin) [inline vars/npmRelease.groovy]: email and username is getting hardcoded here?
if yes then `flow.setupGitSSH()` could take email id and username params

Comment (rupalibehera) [reply vars/npmRelease.groovy]: I think so `username` and `email_id` are same everywhere hence this may not break anything

Comment (jaseemabid) [reply vars/npmRelease.groovy]: @hrishin Its OK. We were not overriding it anywhere. Now things are at least consistent everywhere.

File: vars/goRelease.groovy (hunk 1 above)
Comment (jaseemabid) [reply vars/goRelease.groovy]: @hrishin @rupalibehera I'm not very sure about this one. Should the flow.setupGitSSH() command be inside or outside the go container?

**Comment by jaseemabid** [comment]
@hrishin Feel free to be very picky about this PR because we both know getting this wrong is going to break several builds downstream :D

**Comment by jaseemabid** [comment]
@rupalibehera Yep! Plan is to test it on at least a few planner repos before merging.

**Comment by jaseemabid** [comment]
Patches sent, each needs to be reverted if it gets merged. 

- [ ] https://github.com/fabric8-ui/fabric8-planner/pull/2519
- [ ] https://github.com/fabric8-ui/fabric8-npm-dependencies/pull/629

**Comment by rupalibehera** [comment]
> Patches sent, each needs to be reverted if it gets merged.
> 
>  fabric8-ui/fabric8-ui#2682
>  fabric8-ui/fabric8-planner#2519
>  fabric8-ui/fabric8-npm-dependencies#629

you mean just closing the PR ?

**Comment by jaseemabid** [comment]
> you mean just closing the PR ?

No. We merged some of them, so that we can test the CD case as well. Once we land this PR here, we will have to send another patch reverting all the 3 changes downstream.

**Comment by rupalibehera** [comment]
@jaseemabid , we need to change here too https://github.com/fabric8-services/fabric8-tenant/blob/master/Jenkinsfile#L57

**Comment by jaseemabid** [comment]
@rupalibehera Let's deal with fabric8-tenant separately.

**Comment by joshuawilson** [comment]
After applying the change to Planner, I was **not** able to get a successful build that pushed the version to `fabric8-npm-dependencies`. 

The change to `fabric8-ui` did **not** allow it to build either.

**Comment by jaseemabid** [comment]
[PR Build](https://jenkins.cd.test.fabric8.io/job/fabric8-ui/job/fabric8-ui/view/change-requests/job/PR-2683/2/console) for fabric8-ui/fabric8-ui failed with

```
+ oc new-project fabric8-ui-pr-2683-fabric8-ui
error: open /root/home/.oc/cd.conf: read-only file system
```
:/

**Comment by jaseemabid** [comment]
Rebuilding f8-ui/f8-ui PR pipeline with the k8s config changes. https://jenkins.cd.test.fabric8.io/blue/organizations/jenkins/fabric8-ui%2Ffabric8-ui/detail/PR-2683/3/

**Comment by jaseemabid** [comment]
I had to rebase and rewrite this patch after it caused too many merge conflicts with master. Will test this out tomorrow.

**Comment by jaseemabid** [comment]
Will fix https://github.com/openshiftio/openshift.io/issues/2977 with this PR.

File: vars/deployOpenShiftSnapshot.groovy [hunk 3]
```diff
@@ -43,6 +43,17 @@ def call(body) {
         template = true
     }
     container('clients') {
+        flow.setupK8sConfig()
+
+        sh "cat /root/home/.oc/cd.conf"
+
+        echo "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
```
Comment (sthaha) [inline vars/deployOpenShiftSnapshot.groovy]: debug statement?

File: vars/deployOpenShiftSnapshot.groovy [hunk 4]
```diff
@@ -43,6 +43,17 @@ def call(body) {
         template = true
     }
     container('clients') {
+        flow.setupK8sConfig()
+
+        sh "cat /root/home/.oc/cd.conf"
```
Comment (sthaha) [inline vars/deployOpenShiftSnapshot.groovy]: wouldn't this print the token that is supposed to be a secret?

File: vars/promoteArtifacts.groovy [hunk 5]
```diff
@@ -7,15 +7,14 @@ def call(body) {
...
     def containerName = config.containerName ?: 'maven'
 
     container(name: containerName) {
-        sh 'chmod 600 /root/.ssh-git/ssh-key'
-        sh 'chmod 600 /root/.ssh-git/ssh-key.pub'
-        sh 'chmod 700 /root/.ssh-git'
+        flow.setupGitSSH()
 
         echo "About to release ${name} repo ids ${repoIds}"
         def flow = new io.fabric8.Fabric8Commands()
```
Comment (sthaha) [inline vars/promoteArtifacts.groovy]: isn't this shadowing an earlier declaration? if flow ins't needed outside the block, why not move it close its usage?
There could be similar errors else where

File: vars/pushDockerfileEnvVarChangePR.groovy [hunk 6]
```diff
@@ -41,13 +36,7 @@ def call(body) {
...
 
         container(name: containerName) {
-
-            sh 'chmod 600 /root/.ssh-git/ssh-key'
-            sh 'chmod 600 /root/.ssh-git/ssh-key.pub'
-            sh 'chmod 700 /root/.ssh-git'
-
-            sh "git config --global user.email fabric8-admin@googlegroups.com"
-            sh "git config --global user.name fabric8-release"
+            flow.setupGitSSH()
```
Comment (sthaha) [inline vars/pushDockerfileEnvVarChangePR.groovy]: may be worth moving `def flow` close to its usage.

Comment (jaseemabid) [reply vars/pushDockerfileEnvVarChangePR.groovy]: flow is used in several places in this file and hence defined at top.

**Comment by jaseemabid** [comment]
@sthaha Thanks for the review :tada: :tada: 

1. The debug statements - I've to run this a few times today and hence left it there for testing. All of them will get removed before merge. There is still a DNM tag. 

2. Regarding breaking down this into smaller patches - Its quite painful to do large refactor on this project manually and hence I'm relying on intelllij for code insights, formatting etc. You see patch noise now because the code is so bad and there is so much an IDE can do. This will eventually stop, but now its painful and not worth it to break it down to smaller atomic commits. 

3. Shadowing flow, I did a quick grep and looks like there isn't any more of it. Thanks for spotting!

**Comment by sthaha** [comment]
@jaseemabid , 
1. my bad; missed the DNM flag
2. sigh ... lets go ahead with the monolithic patch then (pragmatic)
3. you are welcome :)

**Comment by jaseemabid** [comment]
Reopened the dummy PR at https://github.com/fabric8-ui/fabric8-ui/pull/2692, rebuilding at https://jenkins.cd.test.fabric8.io/blue/organizations/jenkins/fabric8-ui%2Ffabric8-ui/detail/PR-2692/1/pipeline

**Comment by jaseemabid** [comment]
And a master build https://jenkins.cd.test.fabric8.io/job/fabric8-ui/job/fabric8-ui/job/master/463/console

File: src/io/fabric8/Fabric8Commands.groovy [hunk 7]
```diff
@@ -988,8 +988,8 @@ def openShiftImageStreamInstall(String name, String location) {
 @NonCPS
 def setupGitSSH() {
     sh """
-       git config user.email fabric8-admin@googlegroups.com
-       git config user.name fabric8-release
+       git config --global user.email fabric8-admin@googlegroups.com
```
Comment (sthaha) [inline src/io/fabric8/Fabric8Commands.groovy]: @rupalibehera could you please elaborate (as it is missing in the commit message) why we want to set the config globally?

Comment (jaseemabid) [reply src/io/fabric8/Fabric8Commands.groovy]: @sthaha I've seen both global and local in the previous code, this is just making the sore thumb standout now. This code is slightly safer to execute, like before a git clone.

@rupalibehera Ideally add a 'why' as well to the commit message.

Comment (rupalibehera) [reply src/io/fabric8/Fabric8Commands.groovy]: below is the error mesaage which I was seeing, hence the above change 
```

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'root@jenkins-slave-pf6x7-83gmp.(none)')
```
https://jenkins.cd.test.fabric8.io/job/fabric8-ui/job/fabric8-ui/job/master/513/console

**Comment by rupalibehera** [comment]
One the above PR merges we can safely merge this too https://github.com/fabric8-jenkins/jenkins-openshift-base/pull/25

**Comment by piyush1594** [comment]
I checked this fix by releasing model and got success... boom :tada:

**Comment by jaseemabid** [comment]
@rupalibehera @piyush1594 and I did some tests and a bunch of dependent projects are building well now. This patch can be merged now after one more careful review. :tada:

**Comment by jaseemabid** [comment]
Merging this. +1 from @rupalibehera and @piyush1594

**Comment by jaseemabid** [comment]
Closing https://github.com/openshiftio/openshift.io/issues/3324

**Pull request merged**